File: packages/python/openrouter_auto/storage.py
"""
OpenRouter Auto - Storage Adapters
Multiple storage options: memory, file
"""

import json
import os
from typing import Optional, Dict, Any, TypeVar, Generic
from abc import ABC, abstractmethod
from pathlib import Path

T = TypeVar("T")


class StorageAdapter(ABC):
    """Abstract base class for storage adapters"""

    @abstractmethod
    async def get(self, key: str) -> Optional[Any]:
        """Get value by key"""
        pass

    @abstractmethod
    async def set(self, key: str, value: Any) -> None:
        """Set value by key"""
        pass

    @abstractmethod
    async def remove(self, key: str) -> None:
        """Remove value by key"""
        pass

    @abstractmethod
    async def clear(self) -> None:
        """Clear all values"""
        pass


class MemoryStorage(StorageAdapter):
    """In-memory storage adapter"""

    def __init__(self):
        self._store: Dict[str, Any] = {}

    async def get(self, key: str) -> Optional[Any]:
        return self._store.get(key)

    async def set(self, key: str, value: Any) -> None:
        self._store[key] = value

    async def remove(self, key: str) -> None:
        self._store.pop(key, None)

    async def clear(self) -> None:
        self._store.clear()


class FileStorage(StorageAdapter):
    """File-based storage adapter"""

    def __init__(self, config_path: str = "./.openrouter-auto.json"):
        resolved = Path(config_path).resolve()
        # Guard against path traversal — config must live under CWD or user home
        allowed_roots = [Path.cwd().resolve(), Path.home().resolve()]
        if not any(str(resolved).startswith(str(root)) for root in allowed_roots):
            raise ValueError(
                f"configPath '{config_path}' resolves outside CWD and HOME. "
                "This is not allowed to prevent path traversal."
            )
        self.config_path = resolved
        self._data: Dict[str, Any] = {}
        self._load()

    def _load(self) -> None:
        """Load data from file"""
        try:
            if self.config_path.exists():
                with open(self.config_path, "r", encoding="utf-8") as f:
                    self._data = json.load(f)
        except Exception as e:
            print(f"Error loading config file: {e}")
            self._data = {}

    def _save(self) -> None:
        """Save data to file"""
        import stat
        try:
            self.config_path.parent.mkdir(parents=True, exist_ok=True)
            with open(self.config_path, "w", encoding="utf-8") as f:
                json.dump(self._data, f, indent=2, default=str)
            # Restrict to owner read/write only
            self.config_path.chmod(stat.S_IRUSR | stat.S_IWUSR)
        except Exception as e:
            print(f"Error saving config file: {e}")

    async def get(self, key: str) -> Optional[Any]:
        return self._data.get(key)

    async def set(self, key: str, value: Any) -> None:
        self._data[key] = value
        self._save()

    async def remove(self, key: str) -> None:
        if key in self._data:
            del self._data[key]
            self._save()

    async def clear(self) -> None:
        self._data.clear()
        self._save()

    def get_config(self) -> Dict[str, Any]:
        """Get full config"""
        return dict(self._data)

    def set_config(self, config: Dict[str, Any]) -> None:
        """Set full config"""
        self._data = dict(config)
        self._save()


def create_storage(storage_type: str, config_path: Optional[str] = None) -> StorageAdapter:
    """Create storage adapter by type"""
    if storage_type == "file":
        return FileStorage(config_path or "./.openrouter-auto.json")
    elif storage_type == "memory":
        return MemoryStorage()
    else:
        raise ValueError(f"Unknown storage type: {storage_type}")


# Storage keys
STORAGE_KEYS = {
    "MODELS": "models",
    "MODEL_CONFIGS": "model_configs",
    "USER_PREFERENCES": "user_preferences",
    "LAST_FETCH": "last_fetch",
}


async def get_stored_models(storage: StorageAdapter) -> list:
    """Get stored models"""
    return await storage.get(STORAGE_KEYS["MODELS"]) or []


async def set_stored_models(storage: StorageAdapter, models: list) -> None:
    """Set stored models"""
    await storage.set(STORAGE_KEYS["MODELS"], models)


async def get_model_configs(storage: StorageAdapter) -> Dict[str, Any]:
    """Get model configs"""
    return await storage.get(STORAGE_KEYS["MODEL_CONFIGS"]) or {}


async def set_model_config(storage: StorageAdapter, model_id: str, config: Any) -> None:
    """Set model config"""
    from .types import ModelConfig

    configs = await get_model_configs(storage)
    if isinstance(config, ModelConfig):
        configs[model_id] = config.to_dict()
    else:
        configs[model_id] = config
    await storage.set(STORAGE_KEYS["MODEL_CONFIGS"], configs)


async def remove_model_config(storage: StorageAdapter, model_id: str) -> None:
    """Remove model config"""
    configs = await get_model_configs(storage)
    if model_id in configs:
        del configs[model_id]
        await storage.set(STORAGE_KEYS["MODEL_CONFIGS"], configs)


async def get_user_preferences(storage: StorageAdapter) -> Optional[Any]:
    """Get user preferences"""
    return await storage.get(STORAGE_KEYS["USER_PREFERENCES"])


async def set_user_preferences(storage: StorageAdapter, preferences: Any) -> None:
    """Set user preferences — strips api_key before persisting."""
    safe = preferences
    if hasattr(preferences, '__dict__'):
        safe_dict = {k: v for k, v in preferences.__dict__.items() if k not in ("api_key", "apiKey")}
        safe = safe_dict
    elif isinstance(preferences, dict):
        safe = {k: v for k, v in preferences.items() if k not in ("api_key", "apiKey")}
    await storage.set(STORAGE_KEYS["USER_PREFERENCES"], safe)


async def get_last_fetch_time(storage: StorageAdapter) -> Optional[int]:
    """Get last fetch time"""
    return await storage.get(STORAGE_KEYS["LAST_FETCH"])


async def set_last_fetch_time(storage: StorageAdapter, timestamp: int) -> None:
    """Set last fetch time"""
    await storage.set(STORAGE_KEYS["LAST_FETCH"], timestamp)

File: packages/core/src/storage.ts
/**
 * OpenRouter Auto - Storage Adapters
 * Multiple storage options: memory, localStorage, file, custom
 */

import { StorageAdapter, StorageType, UserPreferences, ModelConfig } from './types';

// In-memory storage (default for Node.js)
class MemoryStorage implements StorageAdapter {
  private store: Map<string, any> = new Map();

  async get<T>(key: string): Promise<T | null> {
    return this.store.get(key) || null;
  }

  async set<T>(key: string, value: T): Promise<void> {
    this.store.set(key, value);
  }

  async remove(key: string): Promise<void> {
    this.store.delete(key);
  }

  async clear(): Promise<void> {
    this.store.clear();
  }
}

// Keys that must never be persisted to browser localStorage
const SENSITIVE_KEYS = ['apiKey', 'api_key'];

// LocalStorage adapter (for browser)
class LocalStorageAdapter implements StorageAdapter {
  private prefix: string = 'ora_';

  async get<T>(key: string): Promise<T | null> {
    if (typeof window === 'undefined' || !window.localStorage) {
      return null;
    }
    try {
      const item = window.localStorage.getItem(this.prefix + key);
      return item ? JSON.parse(item) : null;
    } catch (error) {
      console.error('Error reading from localStorage:', error);
      return null;
    }
  }

  async set<T>(key: string, value: T): Promise<void> {
    if (typeof window === 'undefined' || !window.localStorage) {
      return;
    }
    try {
      // Strip sensitive fields before writing to localStorage
      let safeValue = value;
      if (value && typeof value === 'object' && !Array.isArray(value)) {
        const obj = { ...(value as Record<string, unknown>) };
        for (const k of SENSITIVE_KEYS) {
          delete obj[k];
        }
        safeValue = obj as unknown as T;
      }
      window.localStorage.setItem(this.prefix + key, JSON.stringify(safeValue));
    } catch (error) {
      console.error('Error writing to localStorage:', error);
    }
  }

  async remove(key: string): Promise<void> {
    if (typeof window === 'undefined' || !window.localStorage) {
      return;
    }
    try {
      window.localStorage.removeItem(this.prefix + key);
    } catch (error) {
      console.error('Error removing from localStorage:', error);
    }
  }

  async clear(): Promise<void> {
    if (typeof window === 'undefined' || !window.localStorage) {
      return;
    }
    try {
      const keys: string[] = [];
      for (let i = 0; i < window.localStorage.length; i++) {
        const key = window.localStorage.key(i);
        if (key?.startsWith(this.prefix)) {
          keys.push(key);
        }
      }
      keys.forEach(key => window.localStorage.removeItem(key));
    } catch (error) {
      console.error('Error clearing localStorage:', error);
    }
  }
}

// File storage adapter (for Node.js with config file)
class FileStorageAdapter implements StorageAdapter {
  private configPath: string;
  private data: Record<string, any> = {};
  private fs: any;
  private path: any;

  constructor(configPath: string = './.openrouter-auto.json') {
    this.configPath = configPath;
    
    // Dynamic import for Node.js modules
    try {
      this.fs = require('fs');
      this.path = require('path');
      this.loadFromFile();
    } catch (error) {
      console.warn('File storage not available in browser environment');
    }
  }

  private loadFromFile(): void {
    if (!this.fs) return;
    
    try {
      if (this.fs.existsSync(this.configPath)) {
        const content = this.fs.readFileSync(this.configPath, 'utf-8');
        this.data = JSON.parse(content);
      }
    } catch (error) {
      console.error('Error loading config file:', error);
      this.data = {};
    }
  }

  private saveToFile(): void {
    if (!this.fs) return;
    
    try {
      const dir = this.path.dirname(this.configPath);
      if (!this.fs.existsSync(dir)) {
        this.fs.mkdirSync(dir, { recursive: true });
      }
      this.fs.writeFileSync(
        this.configPath,
        JSON.stringify(this.data, null, 2),
        { encoding: 'utf-8', mode: 0o600 }
      );
    } catch (error) {
      console.error('Error saving config file:', error);
    }
  }

  async get<T>(key: string): Promise<T | null> {
    return this.data[key] || null;
  }

  async set<T>(key: string, value: T): Promise<void> {
    this.data[key] = value;
    this.saveToFile();
  }

  async remove(key: string): Promise<void> {
    delete this.data[key];
    this.saveToFile();
  }

  async clear(): Promise<void> {
    this.data = {};
    this.saveToFile();
  }

  // Get full config object
  getConfig(): Record<string, any> {
    return { ...this.data };
  }

  // Set full config object
  setConfig(config: Record<string, any>): void {
    this.data = { ...config };
    this.saveToFile();
  }
}

// Storage factory
export function createStorage(type: StorageType, configPath?: string): StorageAdapter {
  switch (type) {
    case 'localStorage':
      return new LocalStorageAdapter();
    case 'file':
      return new FileStorageAdapter(configPath);
    case 'memory':
    default:
      return new MemoryStorage();
  }
}

// Storage keys
export const STORAGE_KEYS = {
  MODELS: 'models',
  MODEL_CONFIGS: 'model_configs',
  USER_PREFERENCES: 'user_preferences',
  LAST_FETCH: 'last_fetch',
  MODEL_CACHE: 'model_cache',
} as const;

// Helper functions for common storage operations
export async function getStoredModels(storage: StorageAdapter): Promise<any[]> {
  return (await storage.get(STORAGE_KEYS.MODELS)) || [];
}

export async function setStoredModels(storage: StorageAdapter, models: any[]): Promise<void> {
  return storage.set(STORAGE_KEYS.MODELS, models);
}

export async function getModelConfigs(storage: StorageAdapter): Promise<Record<string, ModelConfig>> {
  return (await storage.get(STORAGE_KEYS.MODEL_CONFIGS)) || {};
}

export async function setModelConfig(
  storage: StorageAdapter,
  modelId: string,
  config: ModelConfig
): Promise<void> {
  const configs = await getModelConfigs(storage);
  configs[modelId] = config;
  return storage.set(STORAGE_KEYS.MODEL_CONFIGS, configs);
}

export async function removeModelConfig(storage: StorageAdapter, modelId: string): Promise<void> {
  const configs = await getModelConfigs(storage);
  delete configs[modelId];
  return storage.set(STORAGE_KEYS.MODEL_CONFIGS, configs);
}

export async function getUserPreferences(storage: StorageAdapter): Promise<UserPreferences | null> {
  return storage.get(STORAGE_KEYS.USER_PREFERENCES);
}

export async function setUserPreferences(
  storage: StorageAdapter,
  preferences: UserPreferences
): Promise<void> {
  return storage.set(STORAGE_KEYS.USER_PREFERENCES, preferences);
}

export async function getLastFetchTime(storage: StorageAdapter): Promise<number | null> {
  return storage.get(STORAGE_KEYS.LAST_FETCH);
}

export async function setLastFetchTime(storage: StorageAdapter, timestamp: number): Promise<void> {
  return storage.set(STORAGE_KEYS.LAST_FETCH, timestamp);
}

export { MemoryStorage, LocalStorageAdapter, FileStorageAdapter };
